### ISSUE TITLE
[Ray[Data]] Put failure: FileNotFoundError

### ISSUE THREAD

**Issue by frazierprime** [ray-project/ray#43282]
### What happened + What you expected to happen

I am trying to write a function for preparing and "putting" dataset on a remote data cluster, but I am getting an error which does not make much sense.

The error references a file that very clearly exists on my jupyter machine:
```
FileNotFoundError: [Errno 2] Failed to open local file '/home/jovyan/.cache/huggingface/datasets/csv/default-f41220f1f9944b95/0.0.0/eea64c71ca8b46dd3f537ed218fc9bf495d5707789152eb2764f5c78fa66d59d/csv-train.arrow'. Detail: [errno 2] No such file or directory
```

See attached screenshots that prove the file in question exists.
<img width="982" alt="Screenshot 2024-02-20 at 11 32 50 AM" src="https://github.com/ray-project/ray/assets/20911937/0463b17a-673a-4f35-9316-68c0eb77e7e0">

<img width="1006" alt="2024-02-20_11-33-59" src="https://github.com/ray-project/ray/assets/20911937/889fb6be-cb8b-4a18-90de-f083cd2f51c8">

File referenced in the FileNotFound error:
```
/home/jovyan/.cache/huggingface/datasets/csv/default-f41220f1f9944b95/0.0.0/eea64c71ca8b46dd3f537ed218fc9bf495d5707789152eb2764f5c78fa66d59d/csv-train.arrow
```

File found on local disk:
```
/home/jovyan/.cache/huggingface/datasets/csv/default-f41220f1f9944b95/0.0.0/eea64c71ca8b46dd3f537ed218fc9bf495d5707789152eb2764f5c78fa66d59d/csv-train.arrow
```

### Versions / Dependencies

Ray SDK version: 2.9.2
Ray cluster version: 2.9.0

### Reproduction script

We are using a dataset that was curated by our team, not an off the shelf dataset from Huggingface, but it adheres to the Dataset object semantics, and I was able to split it using the train_test_split api:

```
# create dataset:
dataset = load_dataset("csv", data_files=PROCESSED_QA_DATA_PATH)

dataset
DatasetDict({
    train: Dataset({
        features: ['instruction', 'input', 'output', 'text'],
        num_rows: 1632
    })
})

# Split on "train" key
train_test_dataset = dataset["train"].train_test_split(test_size=0.2)

# Assign values to train and test datasets.
train_dataset = train_test_dataset["train"]
test_dataset = train_test_dataset["test"]

# Create remote function task.
import ray.data

@ray.remote
def ray_dataset_task():
    ray_datasets = {"train": ray.data.from_huggingface(train_dataset), 
                    "test": ray.data.from_huggingface(test_dataset),}
    return ray_datasets

# This line is the one that throws the error:
ray.get(ray_dataset_task.remote())

### Issue Severity

High: It blocks me from completing my task.

**Comment by frazierprime** [comment]
I tried originally tried calling the `from_huggingface` function without the `remote` wrapper, but I get an error;
```
RuntimeError: Global node is not initialized.
```

I found this ticket, where your team indicates that it should be wrapped in a `ray.remote` context 
https://github.com/ray-project/ray/issues/41333

**Comment by frazierprime** [comment]
Another data point:
I was able to create datasets (without the remote wrapper) by reverting my local Ray version to `2.7.0`, but now I cannot run a training job against the cluster because the versions do not match.
Cluster: 2.9.1
Local SDK: 2.7.0

This is the error I am observing: https://stackoverflow.com/questions/75403232/attributeerror-cant-get-attribute-tunerinternal-validate-overwrite-trainable

**Comment by bveeramani** [comment]
Hey @frazierprime, are you running this program on a multi-node cluster? If the file exists on the head node and you create the dataset in a remote task, then the remote task might be scheduled on a worker node and the file will be inaccessible.

Re: the global node not initialized thing -- we don't really support Ray Client. If possible, I'd recommend not using it.

**Comment by frazierprime** [comment]
> Hey @frazierprime, are you running this program on a multi-node cluster? If the file exists on the head node and you create the dataset in a remote task, then the remote task might be scheduled on a worker node and the file will be inaccessible.
> 
> Re: the global node not initialized thing -- we don't really support Ray Client. If possible, I'd recommend not using it.

Yes, I am testing in a multi-node cluster. Must've missed it in the docs, but how can I ensure that files are accessible to all worker nodes? 

Same question for local packages that I have pushed in with my runtime environment? I am running into a similar issue with this mechanism where I am able to push a local module using the runtime environment, but when the code gets executed on the workers, I get module not found errors.

**Comment by frazierprime** [comment]
@bveeramani - relating to your comment about Ray Client; has this been deprecated? We're using Ray version 2.7.0 in our remote clusters due to this particular issue (I am able to run jobs successfully with the older version).

**Comment by bveeramani** [comment]
My understanding is that it's not explicitly deprecated, but we haven't maintained it for a very long time.

**Comment by bveeramani** [comment]
> Yes, I am testing in a multi-node cluster. Must've missed it in the docs, but how can I ensure that files are accessible to all worker nodes?

In general, you use a distributed filesystem like NFS (or cloud storage if you're reading files directly). For Ray Client, though, I'm not sure if I have any concrete recommendations.

> I am running into a similar issue with this mechanism where I am able to push a local module using the runtime environment, but when the code gets executed on the workers, I get module not found errors.

That's weird. If you specify modules in your runtime environment, I'd expect them to be available on your worker nodes. @rynewang do you have any hunches about what might be happening?
